Implement Format-Hex in C#

MiaRomero opened the pull request:
Addressing issue https://github.com/PowerShell/PowerShell/issues/2730 'Rewrite Format-Hex in C#'

-  [Documentation](https://msdn.microsoft.com/en-us/powershell/reference/5.1/microsoft.powershell.utility/format-hex) for Format-Hex states both the Path and LiteralPath parameters accept string arrays, though this functionality was not actually present for the cmdlet.  This has been corrected so both parameters can accept and process an array of strings.  If an invalid path is included in the array, a non-terminating 'FileNotFound' error will be thrown, but the remaining valid paths will be processed and output as usual.

-  After a discussion with @Francisco-Gamino and @joeyaiello, it was decided to make the Raw parameter No-op.  Going forward all of the output will be displayed with a true representation of numbers that includes all of the bytes for its type (what the Raw parameter was formally doing).  For example, [Int32]170 passed to Format-Hex would have previously displayed only "AA".  Now it will display "AA 00 00 00".

- Updated the InputObject parameter to accept an Int32[], in addition to all previously accepted types.

- The way Format-Hex processes and displays input from the pipeline has been corrected.  Previously, the output for numbers in an array or range would be combined into one as seen below:
```
  PS D:\GitHub\PowerShell> 1,2 | fhx


           Path:

           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F

  00000000   01 02                                            ..
```

 Now each number will be displayed as its own input object:
```
  PS D:\GitHub\PowerShell> 1,2 |fhx


           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F

  00000000   01 00 00 00                                      ....


           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F

  00000000   02 00 00 00                                      ....
```

- Added a new error that states Format-Hex only supports FileSystem Provider Paths.  Previously an error was thrown, but was not specific about the problem (and was displaying the wrong path).
```
fhx : The given path 'Cert:\CurrentUser\My\' is not supported. This command only supports the FileSystem Provider paths.
At line:1 char:1
+ fhx -path Cert:\CurrentUser\My\
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (Cert:\CurrentUser\My\:String) [Format-Hex], ArgumentException
    + FullyQualifiedErrorId : FormatHexOnlySupportsFileSystemPaths,Microsoft.PowerShell.Commands.FormatHex
```

- Cleaned up how we are exposing the Format-Hex error messages.  Removed from UtilityCommon.cs and are being accessed through UtilityCommonStrings.resx instead.

- Updated and added tests for this cmdlet to validate newly implemented functionality and to be more comprehensive for original functionality.

Commit:
Implement Format-Hex in C#

Commit:
Update Format-Hex test cases and add new tests for added functionality

msftclas commented:
@__MiaRomero__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

Francisco-Gamino commented:
@SandeepSutari, @mirichmo, @daxian-dbw, @joeyaiello: Could you please take a look? Thanks.

SandeepSutari commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+        /// </summary>
+        /// <param name="encoding"></param>
+        /// <returns></returns>
+        private Encoding GetEncoding(string encoding)
```
Why is the private member not camel cased ? Unless Pascal casing is the recommended pattern for C# cmdlets, we should use camel casing

SandeepSutari replied:
I mean method name...var name is fine

MiaRomero replied:
I talked to @daxian-dbw about this, and Pascal casing is the recommended pattern for the cmdlets.

SandeepSutari commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+            {
+                string inputObject = obj.ToString();
+                byte[] inputBytes = GetEncoding(_encoding).GetBytes(inputObject);
+                ConvertToHexidecimal(inputBytes, null, 0);
```
Looks like you could get the inputBytes out of each of these else if {} blocks and do the convert* at the end if else block. that would prevent 10 uses of convert calls peppered across this method.

MiaRomero replied:
Fixed.

SandeepSutari commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+        /// that array on to the ConvertToHexidecimal method to output.
+        /// </summary>
+        /// <param name="_inputObject"></param>
+        private void ProcessObjectContent(PSObject _inputObject)
```
unless this is the pattern or a standard practice to use _ for private members we should avoid it. Also why is the method variable using the private member style...we should stick with camel casing since its a method variable.

daxian-dbw replied:
parameter names shouldn't have `'_'` prefixed to them.

MiaRomero replied:
Fixed.

SandeepSutari commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+            else
+            {
+                string errorMessage = StringUtil.Format(UtilityCommonStrings.FormatHexTypeNotSupported, obj.GetType());
+                ErrorRecord errorRecord = new ErrorRecord(new ArgumentException(errorMessage),
```
I would fix indentation here

MiaRomero replied:
Fixed for all error messages.

SandeepSutari commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+            {
+                List<byte> inputStreamArray = new List<byte>();
+                Int64[] inputInts = (Int64[])obj;
+                foreach (Int64 i64 in inputInts)
```
i64 should get itself a better var name. currently its a type name.

MiaRomero replied:
Fixed.

SandeepSutari commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+        /// <returns></returns>
+        private Encoding GetEncoding(string encoding)
+        {
+            if (String.Equals(encoding, "BigEndianUnicode", StringComparison.OrdinalIgnoreCase))
```
use string constants instead of raw strings

MiaRomero replied:
Fixed.

SandeepSutari reviewed (changes requested):
here are a few high level comments

SteveL-MSFT commented on githublog.log:
```diff
@@ -0,0 +1,1225 @@
+﻿**********************
```
this file shouldn't be part of the PR

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+        [ValidateNotNullOrEmpty()]
+        public string[] Path 
+        { 
+            get { return _path; }
```
convention is to use automatic properties:

```c#
public string[] Path { get; set; }
```

MiaRomero replied:
Fixed.

iSazonov commented:
Add `Breaking-Change` because `-Raw` removed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
+
+        It "Validate the alias for Format-Hex cmdlet 'Get-Command fhx -ErrorAction Stop'" {
+
+            try
```
This test case could probably be simplified to:

{ Get-Alias fhx } | Should Not Throw

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
+            $result | Should Not BeNullOrEmpty
+            $result.Count | should be 2
+            $result[0].ToString() | Should be "00000000   54 68 69 73 20 69 73 20 61 20 62 69 74 20 6D 6F  This is a bit mo"
+            if ( $IsCoreCLR ) {
```
Why is the output different on CoreCLR?

MiaRomero replied:
It's not - removed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
-            $result.CommandType.ToString() | Should Be "Alias"
+            ,$result | Should BeOfType 'Microsoft.PowerShell.Commands.ByteCollection'
+            $actualResult = $result.ToString() -split "`r`n"
+            $actualResult.Count | should be 2
```
I am not sure if you need this check here since you are already checking both results below, e.g., $actualResult[0], $actualResult[1]

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
+            $result | Should Not BeNullOrEmpty
+            ,$result | Should BeOfType 'Microsoft.PowerShell.Commands.ByteCollection'
+            $actualResult = $result.ToString()
+            ($actualResult -match $inputText1) | Should Be $true
```
This should be '$actualResult | Should Match $inputText1'; otherwise, when the test fails, it going to say false not equals true.

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
+            $result | Should Not BeNullOrEmpty
+            ,$result | Should BeOfType 'Microsoft.PowerShell.Commands.ByteCollection'
+            $actualResult = $result.ToString()
+            ($actualResult -match $inputText1) | Should Be $true
```
This should be '$actualResult | Should Match $inputText1'; otherwise, when the test fails, it going to say false not equals true.

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
-    It "ValidateInteger64" {
+            $invalidPath = "$($inputFile1.DirectoryName)\fakefile8888845345345348709.txt"
+            $invalid2 = "C:\Windows\fakefile99342039847.txt"
+            $a = $null
```
Could you please use a more meaningful variable name? E.g., $output, and $errorThrown?

MiaRomero replied:
Fixed.

msftclas commented:
__@MiaRomero__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
+                InputObjectErrorCase = $true
+                Path = $inputFile1
+                InputObject = @{ "hash" = "table" }
+                expectedError = "FormatHexTypeNotSupported,Microsoft.PowerShell.Commands.FormatHex"
```
'expectedError' should be rename to 'expectedFullyQualifiedErrorId'.

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
-        $actualResult | should be "00000000   41 42 43 44                                      ABCD            "
+    Context "Correct Expected Errors" {
+
+        $otherProvider = Get-ChildItem Cert:\CurrentUser\My\
```
$otherProvider should probably be renamed to '$certificateProvider'.

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
-        $actualResult = $result.ToString()
-        # whitespace sensitive
-        $actualResult | should be "00000000   41 42 43 44                                      ABCD            "
+    Context "Correct Expected Errors" {
```
Could you please rename this Context to something like "Validate Format-Hex error scenarios" or "Validate negative test cases" or "Validate error scenarios"?

MiaRomero replied:
Fixed.

Francisco-Gamino commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
+                Name = "Can process int[] type '[int[]](5,6) | fhx'"
+                inputObject = [int[]](5,6)
+                Count = 2
+                Expected = "00000000   05 00 00 00                                      ...."
```
Should these be ExpectedResult, and ExpectedSecondResult?

MiaRomero replied:
Fixed.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+    /// <summary>
+    /// Displays the hexidecimal equivalent of the input data.
+    /// <summary>
+    [Cmdlet(VerbsCommon.Format, "Hex", SupportsShouldProcess = true, HelpUri ="https://go.microsoft.com/fwlink/?LinkId=821773")]
```
Shouldn't this be `https`?

MiaRomero replied:
Fixed.

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+        /// Type of character encoding for InputObject
+        /// </summary>
+        [Parameter(ParameterSetName = "ByInputObject")]
+        [ValidateSet("Ascii", "UTF32", "UTF7", "UTF8", "BigEndianUnicode", "Unicode")]
```
yes, this should probably be an enum rather than the string set

---
In reply to: [105936094](https://github.com/PowerShell/PowerShell/pull/3320#discussion_r105936094) [](ancestors = 105936094)

iSazonov replied:
We have [FileSystemCmdletProviderEncoding](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L7329) enum (but in [EncodingConversion](https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/src/System.Management.Automation/utils/PathUtils.cs#L440) the same is string constants in the class. 
It seems we need to bring this to the common code base.)

MiaRomero replied:
Fixed.

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+                foreach (string currentPath in path)
+                {
+                    string newPath = Context.SessionState.Path.GetUnresolvedProviderPathFromPSPath(currentPath);
+                    pathsToProcess.Add(newPath);
```
do you need to validate that these paths are filesystem provider paths?
for example, `format-hex -literalpath hklm:/software` should perhaps produce the same error as `format-hex -path hklm:/software`

MiaRomero replied:
Fixed.

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+                        {
+                            pathsToProcess.Add(newPaths[0]);
+                        }
+                    }
```
it seems like it would be a useful scenario to support multiple files here:
`format-hex file*.txt`
unless that's an explicit non-supported scenario

JamesWTruher replied:
yep - looking below, it appears that it just loops through the provided paths, it seems reasonable to support more than one resolved path

---
In reply to: [105987086](https://github.com/PowerShell/PowerShell/pull/3320#discussion_r105987086) [](ancestors = 105987086)

MiaRomero replied:
Fixed.

lzybkr reviewed (commented):
The file `src\Modules\Unix\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1` also needs updating - that should get the tests passing in Travis.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,460 @@
             $result | Should Not BeNullOrEmpty
-            $result.CommandType.ToString() | Should Be "Alias"
+            ,$result | Should BeOfType 'Microsoft.PowerShell.Commands.ByteCollection'
+            $actualResult = $result.ToString() -split "`r`n"
```
Please use platform independent [Environment]::NewLine

MiaRomero replied:
Fixed.

MiaRomero commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,395 @@
+                        {
+                            // Write a non-terminating error message indicating that path specified is not supported.
+                            string errorMessage = StringUtil.Format(UtilityCommonStrings.FormatHexOnlySupportsFileSystemPaths, currentPath);
+                            ErrorRecord errorRecord = new ErrorRecord(new ArgumentException(errorMessage),
```
Fixed.

MiaRomero commented:
Hi @SandeepSutari, @Francisco-Gamino, @SteveL-MSFT, and @JamesWTruher,
Thank you for your feedback.  I have addressed your comments - would you please take another look?
Thanks,
Maria

SteveL-MSFT reviewed (approved):
My comments have been addressed.  Thanks @MiaRomero !

lzybkr commented:
@daxian-dbw - I don't think we want to get in the business of dealing with endian-ness in this cmdlet.

I think of it as more of a memory dumping tool, and it shouldn't necessarily interpret what it's dumping (even though we could).

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,334 @@
+                inputBytes = resolvedEncoding.GetBytes(inputString);
+            }
+            
+            else if (obj is byte)
```
It would be nice to see all primitive types (including enums) and arrays of these types - instead of this random assortment.

I wrote some code that covers much of what I think is useful and it's a similar number of lines of code, see https://gist.github.com/lzybkr/9940b63c8301c31316bdb3ec6305536f

lzybkr replied:
It might also be useful to support any value type, not just primitives, but that is a little harder.

Francisco-Gamino replied:
Hi @lzybkr, Maria leap program assignment/internship ends today, so I will follow-up on this. I have open an issue https://github.com/PowerShell/PowerShell/issues/3358 to track this. Thanks.

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,334 @@
+
+            foreach (string path in pathsToProcess)
+            {
+                ProcessFileContent(path);
```
It doesn't look like there is any file header, so if there are multiple files, it will be difficult to understand the output.

MiaRomero replied:
There is a header for each ByteCollection object, e.g., 

New-Item -Value "foo" -Path foo1.txt -ItemType File
New-Item -Value "foo2" -Path foo2.txt -ItemType File

PS C:\exp> dir foo* | fhx
           Path: C:\exp\foo1.txt
           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000   66 6F 6F                                         foo

           Path: C:\exp\foo2.txt
           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000   66 6F 6F 32                                      foo2

PS C:\exp> fhx -Path .\foo*
           Path: C:\exp\foo1.txt
           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000   66 6F 6F                                         foo

           Path: C:\exp\foo2.txt
           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
00000000   66 6F 6F 32                                      foo2

lzybkr replied:
Great.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and agreed on:
- for numbers, we should prefix with 0 to retain width
- change to formatting (due to storing in property) is a breaking change that just needs to be documented
- if array is passed in, then each element in the array is a separate output object
- FileInfo retains current behavior (special cased to format the content)

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,488 @@
+
+        It "Validate the alias (fhx) for Format-Hex cmdlet is available" {
+
+            { Get-Alias fhx } | Should Not Throw
```
Please remove the test because we already have separate files with tests for aliases.

MiaRomero replied:
Removed.

MiaRomero commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,334 @@
+        {
+            Object obj = inputObject.BaseObject;
+            byte[] inputBytes = null;
+            if (obj is System.IO.FileSystemInfo)
```
Per committee review/comment below, we'll leave as is for now.

MiaRomero commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,334 @@
+        /// <param name="pathsToProcess"></param>
+        private void ProcessPath(List<string> pathsToProcess)
+        {
+            // If there are no paths to proccess, just return.
```
Thank you.  Good point, it turns out we do not need this code.  It does throw an error if the path does not resolve to a file.  

```
PS > fhx -Path "c:\dlfkjasdkfj.txt", "c:\ldkfj.txt"
fhx : Cannot find path 'C:\dlfkjasdkfj.txt' because it does not exist.
At line:1 char:1
+ fhx -Path "c:\dlfkjasdkfj.txt", "c:\ldkfj.txt"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (System.String[]:String[]) [Format-Hex], ItemNotFoundException
    + FullyQualifiedErrorId : FileNotFound,Microsoft.PowerShell.Commands.FormatHex

fhx : Cannot find path 'C:\ldkfj.txt' because it does not exist.
At line:1 char:1
+ fhx -Path "c:\dlfkjasdkfj.txt", "c:\ldkfj.txt"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (System.String[]:String[]) [Format-Hex], ItemNotFoundException
    + FullyQualifiedErrorId : FileNotFound,Microsoft.PowerShell.Commands.FormatHex

lzybkr replied:
And just to confirm - there is no error if you use a wildcard and there are no matches?

rkeithhill commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,334 @@
+    {
+        private const int BUFFERSIZE = 16;
+
+        #region Parameters
```
Is there any chance we could get `Count` and `Offset` parameters?  I could probably live without `Offset` but on the PSCX Format-Hex command, I use `-Count` almost all the time.  That's because I use it to see if a text file has a BOM, in which case I only want to see the first 3 characters.

MiaRomero replied:
Those are good suggestions, but out of scope for this PR.

Francisco-Gamino replied:
@rkeithhill, please open a issue for that suggestion. Thanks.

MiaRomero commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,488 @@
-        ,$result | Should BeOfType 'Microsoft.PowerShell.Commands.ByteCollection'
-        $actualResult = $result.ToString()
-        ($actualResult -match $inputText1) | Should Be $true
+        $errorThrown | Should Not Be $null
```
In this scenario we want to test that we get both an error and output.

MiaRomero commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,488 @@
+        }
+
+        $thrownError | Should Not Be $null
+        $thrownError.FullyQualifiedErrorId | Should Match $testCase.ExpectedFullyQualifiedErrorId
```
I've tried using this but get this error:
The term 'ShouldBeErrorId' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

So for now I'll keep Should Match

MiaRomero commented on test/powershell/Modules/Microsoft.PowerShell.Utility/FormatHex.Tests.ps1:
```diff
@@ -5,219 +5,488 @@
+
+        foreach ($testCase in $testCases)
+        {
+            RunPathAndLiteralPathParameterTestCase $testCase
```
Yes, that is true.  But I'll leave as is for now.

MiaRomero commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -0,0 +1,334 @@
+                            // will just contain multiple references to the same buffer memory space (containing only the
+                            // last bytes of the file read).  Cloning the buffer allows us to pass the values on without
+                            // overwriting previous values.
+                            byte[] bufferClone = (byte[])buffer.Clone();
```
I've changed it to copy the buffer into a new array instead of cloning the buffer.  If this isn't what you had in mind, please file an issue and @Francisco-Gamino  will follow up.  Due to time constraints with my internship, I won't be able to follow up after today.

MiaRomero commented:
Thank you everyone for your feedback.  @lzybkr can you please merge when you have a chance?  My internship is over so if there are any questions/concerns please address them to @Francisco-Gamino.  Thanks!

Francisco-Gamino commented:
It was great working with you @MiaRomero, thank you!

lzybkr reviewed (approved):
I opened https://github.com/PowerShell/PowerShell/issues/3382 so we clean up the test a little, but otherwise looks good.

Francisco-Gamino commented:
Thanks @lzybkr! I will take care of #3382.
